ports/raspberrypi/common-hal/pulseio/PulseIn.c

```diff
@@ -0,0 +1,242 @@
+        NULL, 0,
+        NULL, 0,
+        1, 0,
+        1 << self->pin, false, true,
+        false, 8, false, // TX, unused
+        false,
+        true, 32, true, // RX auto-push every 32 bits
+        false); // claim pins
+    self->state_machine.pio = state_machine.pio;
+    self->state_machine.state_machine = state_machine.state_machine;
+    self->state_machine.sm_config = state_machine.sm_config;
+    self->state_machine.offset = state_machine.offset;
+    if ( self->state_machine.pio == pio0 ) {
+       self->pio_interrupt = PIO0_IRQ_0;
+    } else {
+       self->pio_interrupt = PIO1_IRQ_0;
+    }
+    pio_sm_clear_fifos(self->state_machine.pio,self->state_machine.state_machine);
+    last_level = self->idle_state;
+    level_count = 0;
+    result = 0;
+    buf_index = 0;
+
+    pio_sm_set_in_pins(state_machine.pio,state_machine.state_machine,pin->number);
+    pio_sm_set_enabled(state_machine.pio,state_machine.state_machine, false);
+    irq_set_exclusive_handler(self->pio_interrupt, common_hal_pulseio_pulsein_interrupt);
+    hw_clear_bits(&state_machine.pio->inte0, 1u << state_machine.state_machine);
+    hw_set_bits(&state_machine.pio->inte0, 1u << (state_machine.state_machine+8));
+
+    // exec a set pindirs to 0 for input
+    pio_sm_exec(state_machine.pio,state_machine.state_machine,0xe080);
+    irq_set_enabled(self->pio_interrupt, true);
+    pio_sm_set_enabled(state_machine.pio, state_machine.state_machine, true);
+}
+
+bool common_hal_pulseio_pulsein_deinited(pulseio_pulsein_obj_t* self) {
+    return self->pin == NO_PIN;
+}
+
+void common_hal_pulseio_pulsein_deinit(pulseio_pulsein_obj_t* self) {
+    if (common_hal_pulseio_pulsein_deinited(self)) {
+        return;
+    }
+    irq_set_enabled(self->pio_interrupt, false);
+    pio_sm_set_enabled(self->state_machine.pio, self->state_machine.state_machine, false);
+    pio_sm_unclaim (self->state_machine.pio, self->state_machine.state_machine);
+    m_free(self->buffer);
+    self->pin = NO_PIN;
+}
+
+void common_hal_pulseio_pulsein_pause(pulseio_pulsein_obj_t* self) {
+    pio_sm_set_enabled(self->state_machine.pio, self->state_machine.state_machine, false);
+}
+
+void common_hal_pulseio_pulsein_interrupt() {
+
+    pulseio_pulsein_obj_t* self = save_self;
+    uint32_t rxfifo = 0;
+
+    rxfifo = pio_sm_get_blocking(self->state_machine.pio, self->state_machine.state_machine);
+    // translate from fifo to buffer
+    for (uint i = 0; i < 32; i++) {
+        bool level = (rxfifo & (1 << i)) >> i;
+        if (level == last_level ) {
+            level_count ++;
+        } else {
+           result = level_count;
+           last_level = level;
+           level_count = 1;
+           // ignore pulses that are too long and too short
+           if (result < 2000 && result > 10) {
+               self->buffer[buf_index] = result;
+               buf_index++;
+               self->len++;
+             }
+           }
+       }
+        gpio_put(pin_GPIO15.number, true);
```

Comment:
```suggestion
```

Reply:
Ooops! Left over from some debugging with the logic analyzer. Removed.